lib/cucumber/formatter/http_io.rb

```diff
@@ -7,86 +8,132 @@ class << self
+        headers = ''
+
+        last_flag = nil
+        chunks.each do |chunk|
+          if ['-X', '--request'].include?(chunk)
+            last_flag = '-X'
+            next
           end
-          write_io.start_request(@http, @req)
 
-          write_io
+          if chunk == '-H'
+            last_flag = '-H'
+            next
+          end
+
+          if last_flag == '-X'
+            http_method = chunk
+            last_flag = nil
+          end
+
+          headers += chunk if last_flag == '-H'
         end
 
-        def build_uri_method_headers(url)
-          uri = URI(url)
-          query_pairs = uri.query ? URI.decode_www_form(uri.query) : []
-
-          # Build headers from query parameters prefixed with http- and extract HTTP method
-          http_query_pairs = query_pairs.select { |pair| pair[0] =~ /^http-/ }
-          http_query_hash_without_prefix = Hash[http_query_pairs.map do |pair|
-                                                  [
-                                                    pair[0][5..-1].downcase, # remove http- prefix
-                                                    pair[1]
-                                                  ]
-                                                end]
-          method = http_query_hash_without_prefix.delete('method') || 'POST'
-          headers = {
-            'transfer-encoding' => 'chunked'
-          }.merge(http_query_hash_without_prefix)
-
-          # Update the query with the http-* parameters removed
-          remaining_query_pairs = query_pairs - http_query_pairs
-          new_query_hash = Hash[remaining_query_pairs]
-          uri.query = URI.encode_www_form(new_query_hash) unless new_query_hash.empty?
-          [uri, method, headers]
+        [
+          url,
+          http_method,
+          make_headers(headers)
+        ]
+      end
+
+      def self.make_headers(headers)
+        hash_headers = {}
+        str_scanner = /("(?<key>[^":]+)\s*:\s*(?<value>[^":]+)")|('(?<key1>[^':]+)\s*:\s*(?<value1>[^':]+)')/
+
+        headers.scan(str_scanner) do |header|
+          header = header.compact!
+          hash_headers[header[0]] = header[1]&.strip
         end
 
-        private
+        hash_headers
+      end
+    end
 
-        def build_request(uri, method, headers)
-          method_class_name = "#{method[0].upcase}#{method[1..-1].downcase}"
-          req = Net::HTTP.const_get(method_class_name).new(uri)
-          headers.each do |header, value|
-            req[header] = value
-          end
-          req
+    class IOHTTPBuffer
+      attr_reader :uri, :method, :headers
+
+      def initialize(uri, method, headers = {}, https_verify_mode = nil)
+        @uri = URI(uri)
+        @method = method
+        @headers = headers
+        @write_io = Tempfile.new('foo')
+
+        @http = build_client(@uri, https_verify_mode)
+      end
+
+      def close
+        @write_io.rewind
+        post_content(@uri, @method, @headers, @write_io.read)
```

Comment:
This reads everything into memory again, so there is little point in using a tempfile if you're going to do that. This is likely to cause out of memory errors when people start attaching a lot of screenshots.

Reply:
You have to investigate 2 possible options:
* Use `IO.copy_stream` and possibly a Thread
* See if a request body can be passed as an IO - I haven't checked

Reply:
Apparently, `Net::Http` does not support IO input, but `http.rb` might. I keep looking (but first I want to get the build passing)

Reply:
I'd rather not add another dependency.

`net/http` does indeed support IO input via `req.body_stream`. In fact it's used on the `master` branch (check the diff of this PR)

Reply:
I'll check that. I though this only worked when treating content (using chunked encoding). I'll have a look.

I agree on not adding extra dependency.

Reply:
You can use `body_stream` without chunked transfer as long as you set `Content-Length`: https://github.com/ruby/ruby/blob/69b3e0ac59f99bbd7e9a76af283f0be1991a3bde/lib/net/http/generic_request.rb#L192

There should be a way to get that from the `Tempfile`...

Reply:
That's done now :)